spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +70,97 @@ def __iter__(self):
+                    #  after we stop looping - https://docs.python.org/3/reference/compound_stmts.html#while
+                    #  This helps us prevent needing to duplicate the self._buffer_memview assignment
+                    self._consumed = True
+            else:
+                self._buffer_memview = memoryview(self._buffer).cast("B").toreadonly()
 
-            # Trim off any data that is past the provided `size` and hold on to it for subsequent `read` calls
-            if len(content) > size:
-                self._trailing_data = content[size:]
-                content = content[:size]
+        start_pos = self._buffer_position
+        self._buffer_position = self._buffer_position + size if (size and not self._consumed) else len(self._buffer)
+        return self._buffer_memview[start_pos:self._buffer_position]
 
-        return content
+    _LINE_CHUNK_SIZE = 1024 * 1024
 
     def iter_lines(self) -> Iterator[bytes]:
         """
         Returns a stream of lines from self._chunks.
         """
-        for chunk in self._chunks:
-            self._add_chunk_to_size(chunk)
-            if self._trailing_data:
-                chunk = self._trailing_data + chunk
-
-            lines = chunk.splitlines()
+        trailing_data = bytearray()
+        # We use a relatively large chunk_size here to minimize the number of times we have to enter this loop/copy data
+        while chunk := self.read(self._LINE_CHUNK_SIZE):
+            if trailing_data:
+                # Instead of appending via e.g. chunk = trailing_data + chunk, which causes a ton of data copying,
+                #  we can just write this chunk into our current line buffer. This is extremely helpful when we
+                #  encounter large files where all the data is actually on one line (i.e. a 500MB JSON file that was
+                #  written without any line breaks). This saves us 10s of seconds when processing these types of files
+                # This is safe to do here because trailing_data is never referenced outside of this function, and we
+                #  only ever return trailing_data when we are done reading in the whole file
+                trailing_data.extend(chunk)
+                chunk = trailing_data
+                trailing_data = bytearray()
+
+            # In the case where we have no trailing_data, then our `chunk` may be a memoryview returned from
+            #  self.read(), which does not support splitlines(), so we need to copy this into a bytearray first
+            bs = chunk if isinstance(chunk, bytearray) else bytearray(chunk)
+
+            # Don't check for `\r\n` because that will be caught by checking for '\n'
+            # Doing this similarly helps keep parsing fast for large single-line files, while having minimal performance
+            #  impact on files that actually have more than 1 line
+            lines = bs.splitlines() if b'\n' in bs else [bs]
```

Comment:
I dropped the check for `b'\r'` since I don't think it's really possible that we're going to be encountering that as a valid line ending.